### ISSUE TITLE
Rule: disable regular expression literal

### ISSUE THREAD

**Issue by LuoZijun** [eslint/eslint#11538]
**Please describe what the rule should do:**

Do not allow Regular Expression Literal.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
(new RegExp("a\bc")).test("a\bc")  // true
/a\bc/.test("a\bc")                // false
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Beacuse this is very important.

**Are you willing to submit a pull request to implement this rule?**

No.

**Comment by aladdin-add** [comment]
can you clarify why this is important?

reminder: it can be achevied using [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax)

**Comment by LuoZijun** [comment]
@aladdin-add  

要解释清楚这个问题，对于我的语言水平来说非常困难，所以这里所有的解释以汉语为准。

To explain this problem is very difficult for my language level, so all the explanations here are based on Chinese.


汉语（Chinese）
------------------

正则表达式实例和正则表达式字面量在当前的实现当中并不总是具有相同的执行结果，这个问题是因为词法分析器(Lexer)导致的。

词法分析器在读取 正则表达式字面量时，遵循的规则是

* [regular-expression-literals](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-literals-regular-expression-literals) 

而在读取 正则表达式实例 时，遵循的规则是
* [string-literals](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-literals-string-literals) ， 先按照字符串的转义规则进行读取
* [regular-expression-objects
](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-regexp-regular-expression-objects
)

从 ECMAScript 9.0 (ES2018) 规范来看，两种词法分析，似乎并没有什么问题。

它们最终的解释行为大概是这样:

Rust Code:
```rust

fn read_regular_expression_literal(source: &str) -> &str;
fn read_string_literal(source: &str) -> &str;
fn parse_regex_pattern(s: &str) -> Pattern;

let source: &str = "ecmascript source code ...";

// Regular Expression Literal
let pattern = parse_regex_pattern(read_string_literal(read_regular_expression_literal(source)));
// Regular Expression String
let pattern = parse_regex_pattern(read_string_literal(source));
```

经过词法分析器得到的 `regular-expression-literals` 其实应该是一个 `RawString`，它依然存在原始的转义字符。所以还需要对这份原始字符串应用 `字符串读取规则（String Literal）`，最终交给 正则对象 解析 `pattern` 。

但是在**现有的解释器实现**当中，第二步显然出现了问题。

所以在各大主流解释器并未修正该问题的情况下，考虑到潜在可能出现的问题，或许我们需要使用 正则表达式实例 来替代 正则表达式字面量  来规避潜在风险。

-- 来自必应翻译


英文(English)
-----------------

Regular expression instances and regular expression literal amounts do not always have the same execution results in the current implementation, this problem is caused by the lexical analyzer (Lexer).

When the lexical parser reads the literal amount of a regular expression, it follows the rules:

* [regular-expression-literals](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-literals-regular-expression-literals) 

When reading a regular expression instance, the rule followed is

* [string-literals](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-literals-string-literals) ， Read first according to the escape rule of the string
* [regular-expression-objects
](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-regexp-regular-expression-objects
)

Judging from the ECMAScript 9.0 (ES2018) specification, there seems to be nothing wrong with the two lexical analyses. Their final act of interpretation is probably this:

Rust Code:
```rust

fn read_regular_expression_literal(source: &str) -> &str;
fn read_string_literal(source: &str) -> &str;
fn parse_regex_pattern(s: &str) -> Pattern;

let source: &str = "ecmascript source code ...";

// Regular Expression Literal
let pattern = parse_regex_pattern(read_string_literal(read_regular_expression_literal(source)));
// Regular Expression String
let pattern = parse_regex_pattern(read_string_literal(source));
```

The ' regular-expression-literals ' obtained by the lexical analyzer should actually be a ' RawString ', it still exists the original escape character.

So you also need to apply the ' string read rule (string Literal) ' to this original string and eventually to the regular object to parse ' pattern '.

However, in ** existing interpreter implementation **, the second step is obviously a problem. So in cases where the major mainstream interpreter does not fix the problem, given the potential problems, perhaps we need to use regular expression instances to override the regular expression literal to avoid potential risk.

-- From Bing Translate

**Comment by LuoZijun** [comment]
@aladdin-add 

对于这个项目，我只是听说过，我从未使用过这个项目。我对JS的生态知道的也很少。
所以对于应该把这个 `规则` 放到 `rules` 里面还是 `no-restricted-syntax` 里面，我其实没有任何想法。

我的初衷是希望有更多的人意识到这个问题。

这个问题是我在写 解释器的 时候发现的。我想 `eslint` 或许可以对这样的语法做出一些提醒。

**Comment by aladdin-add** [comment]
谢谢您的解释，我之前都没注意到这个问题!

我也不清楚 ES spec 是怎样定义这个行为的，但我觉得完全禁止使用正则字面量似乎有些过分，我们已经有一条规则：https://eslint.org/docs/rules/no-control-regex 似乎可以解决这个问题。

thanks for the detail, I would not have noticed it!

I have no idea how ES spec standardize it. But disallowing the usage of all regex literals seems an overkill to me, and we have a rule [no-control-regex](https://eslint.org/docs/rules/no-control-regex) can be useful!

**Comment by g-plane** [comment]
@LuoZijun Example of usage `no-restricted-syntax`. I've not tested it yet so I'm not sure if it's OK/

```json
{
    "rules": {
        "no-restricted-syntax": ["error", "Literal[regex]"]
    }
}
```

**Comment by g-plane** [comment]
含正则表达式的字符串字面量，应该交给专门的正则解析器来解析。以 ESLint 为例，它使用的 ECMAScript parser 是 espree，它不会对字符串字面量作任何解析，而 ESLint 使用 `regexpp` 来完成正则解析工作。

String literals which contains regular expressions should be parsed by a dedicated regular expressions parser. For example, ESLint uses `espree` as the ECMAScript parser, but it won't parse any string literals. Instead, ESLint uses `regexpp` to do this job.

**Comment by g-plane** [comment]
`swc` and `ratel` don't parse string literals as well.

swc:
https://github.com/swc-project/swc/blob/84973890e7424d6ab582c8fe68272a0dc35e9402/ecmascript/parser/src/parser/expr/mod.rs#L272

ratel:
https://github.com/ratel-rust/ratel-core/blob/2e790b33d53755a1cc2d973bf833baebd2f9958b/ratel/src/parser/expression.rs#L429

**Comment by LuoZijun** [comment]
@aladdin-add @g-plane 看起来 `{ "rules": { "no-restricted-syntax": ["error", "Literal[regex]"] } }` 这样子就可以了，这么看来的确不需要额外的规则。

**Comment by LuoZijun** [comment]
@g-plane 

> 含正则表达式的字符串字面量，应该交给专门的正则解析器来解析。以 ESLint 为例，它使用的 ECMAScript parser 是 espree，它不会对字符串字面量作任何解析，而 ESLint 使用 regexpp 来完成正则解析工作。

我觉得重点不在于 `ESLint` 或是类似于 `babel` 这样的库是如何 解析正则语法的，因为这些最终是要在 `Firefox/Chrome/Node.js` 里面执行的，而在当前，这些平台的处理的确存在一些问题。

**Comment by LuoZijun** [comment]
@g-plane 

> `swc` and `ratel` don't parse string literals as well.

这两个库我都贡献过代码，并且提出个很多问题。

`ratel` 是一个非常高性能的语法解析器（词法分析器能做到 900MB+/s 的性能），但是它对ECMAScript语法规范的支持并不完善，并且当前的开发也并不活跃。另外最重要的一点，它能做到如此高的性能，是因为他在很多层面没有验证词法的正确性。同时一些对词法分析器性能有非常大性能损失的特性也未给予支持（比如 Unicode ）。如果它支持了这些特性，实际上它的性能不会超过 100MB/s 。

`swc` 项目看起来是一个经验并不是很丰富的人写的，项目结构复杂，代码较为混乱，性能低下，项目编译性能同样低下（耗时长，产生的二进制可执行程序体积大）。这个项目的 BUG 非常多，实在不推荐使用。另外，在原则上面，这个项目是希望成为一个 `babel` ，但是我并不认可这种，我认为 `babel` 是一个麻烦，是一个错误。我个人不倾向于支持非 ECMAScript 的特性。

所以，目前而言，一个性能好，特性支持完善，完全服从 ECMAScript 规范的 解析器还没有。

这就是为什么有了 [LuoZijun/es](https://github.com/LuoZijun/es/tree/master/src) 这个项目。

**Comment by LuoZijun** [comment]
@g-plane 我看了下你提到的 [eslint/espree](https://github.com/eslint/espree) ，在 Unicode 的处理上面似乎存在一些问题（当然这个问题在 Babel/cherow 里面也同样存在）。

比如 Example Code:
```javascript
var \u{2fa1d} = "\u{10ffff}";   // MAX Unicode CHAR
var \u{2fa1d} = "\u{010ffff}"; // 能容纳最大 6 个 HexDigit， 这里已经超了。
```

**Comment by LuoZijun** [comment]
@g-plane @aladdin-add 

抱歉，前面的我的理解似乎并不对，看起来 @g-plane 的理解才是对的。

正则表达式的字面量相当于其它语言里面的 `RawString ( r" ... " )`, 除了换行符需要明确禁止，所有的转义字符应该都原样保留。然后交给正则表达式解析器解析 `pattern/flags` 。

不过这似乎依然无法解释:

```js
/\b/.test("\b");            // false
/\b/.compile().test("\b");  // true
```

`\b` 字符在正则表达式的转义规则里面也是 `\u{0008}`， 和字符串的转义规则是相同的。

我想这可能只是一个实现问题。

**Comment by g-plane** [comment]
`RegExp.prototype.compile` is deprecated.
See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/compile
